File: typescript/src/Queue.ts
class Queue<T> {
  public data: T[];
  public length: number;

  constructor() {
    this.data = [];
    this.length = 0;
  }

  add(item: T) {
    this.length += 1;
    this.data.push(item);
  }

  remove() {
    if (this.length === 0) {
      return undefined;
    }
    this.length -= 1;
    return this.data.shift();
  }

  peek() {
    return this.data[0];
  }

  last() {
    return this.data[this.length - 1];
  }
}

File: rust/src/data_structures/queue.rs
use std::default::Default;

#[derive(Default)]
#[derive(Debug)]

pub struct Queue<T> {
    pub length: usize,
    pub data: Vec<T>
}

impl<T> Queue<T> {

    pub fn enqueue(&mut self, item: T) {
        self.length += 1;
        self.data.push(item)
    }

    pub fn dequeue(&mut self) -> Option<T> {
        if self.length != 0 {  
            self.length -= 1;
            let item: Option<T> = Some(self.data.remove(0));

            return item;
        }

        None
    }

    pub fn peek(&self) -> Option<&T> {
        return self.data.get(0);
    }
}

pub fn new_queue<T: Default>() -> Queue<T> {
    let queue: Queue<T> = Queue { ..Default::default() };

    queue
}